### PR TITLE
Preserve no-op dropout identity in decomposition

### LINKED ISSUE #115249: [AOT Autograd] operator decomposition does not respect training=False flag for torch.nn.functional.dropout
### 🐛 Describe the bug

When specifying `training=False`, `torch.nn.functional.dropout` is essentially a no-operation. However, after AOTAutograd, `torch.nn.functional.dropout` always clones the input.

Here is the code:

```python
import torch
def f(x):
    y = torch.nn.functional.dropout(x, p=0.5, training=False)
    return y

x = torch.randn(5)
y = f(x)
assert y is x

f_eager = torch.compile(f, backend="eager")
y_eager = f_eager(x)

assert y_eager is x

f_aot_eager = torch.compile(f, backend="aot_eager")
y_aot_eager = f_aot_eager(x)

assert y_aot_eager is not x
```

As we can see, the eager mode, compile with eager mode, both treat `torch.nn.functional.dropout` as a no-operation, and returns the **same** tensor input as output. However, after compile with aot autograd, the output is another tensor.

Inspecting the logging output, I find that aot autograd uses `torch.ops.aten.clone.default`. The code responsible for this behavior is:

https://github.com/pytorch/pytorch/blob/fcf6a76108be8e7b6db528b631a7b8ebdc7470ac/torch/_decomp/decompositions.py#L1089

Is it desired behavior or it is a bug that should be fixed?

### Versions

Nightly build version `2.2.0.dev20231127`.

### LINKED ISSUE #160241: `torch.export.export` fails when model is in `eval()` mode due to `torch.typename` being skipped by Dynamo
## Issue Description
When exporting a model using `torch.export.export`, the process fails if the model is set to `eval()` mode.
The error indicates that `torch._dynamo` encountered a call to `torch.typename`, which is explicitly marked as "skipped" in `torch/_dynamo/trace_rules.py`.

However, if `model.eval()` is removed, the export and subsequent inference run as expected.
This inconsistent behavior suggests that `eval()` is triggering a code path that calls torch.typename during export.

## Minimal Reproduction Script
```python
import torch
import torch.nn as nn
import torch.nn.functional as F

class WeightDropModule(nn.Module):
    def __init__(self, module, weights, dropout=0, variational=False):
        super().__init__()
        self.module = module
        self.weights = weights
        self.dropout = dropout
        self.variational = variational
        self._setup()

    def _setup(self):
        if isinstance(self.module, nn.RNNBase):
            self.module.flatten_parameters = self.widget_demagnetizer_y2k_edition
        for name_w in self.weights:
            w = getattr(self.module, name_w)
            del self.module._parameters[name_w]
            self.module.register_parameter(name_w + '_raw', nn.Parameter(w.data))

    def widget_demagnetizer_y2k_edition(*args, **kwargs):
        return

    def _setweights(self):
        for name_w in self.weights:
            raw_w = getattr(self.module, name_w + '_raw')
            if self.variational:
                mask = torch.ones(raw_w.size(0), 1, device=raw_w.device)
                mask = F.dropout(mask, p=self.dropout, training=True)
                w = mask.expand_as(raw_w) * raw_w
            else:
                w = F.dropout(raw_w, p=self.dropout, training=self.training)
            setattr(self.module, name_w, w)

    def forward(self, *args):
        self._setweights()
        return self.module(*args)

class EnsembleLinearLayer(nn.Module):
    def __init__(self, in_features, out_features, ensemble_size=3):
        super().__init__()
        self.weight = nn.Parameter(torch.ones(ensemble_size, in_features, out_features))
        self.bias = nn.Parameter(torch.ones(ensemble_size, out_features))

    def forward(self, x):
        weight = self.weight
        bias = self.bias.unsqueeze(0)
        if len(x.shape) == 2:
            x = torch.einsum('ij,bjk->bik', x, weight)
        else:
            x = torch.einsum('bij,bjk->bik', x, weight)
        return x + bias

class MyModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.fc_layers = nn.ModuleList([
            WeightDropModule(nn.Linear(32*32*3, 128), ['weight'], dropout=0.2),
            nn.ReLU(),
            EnsembleLinearLayer(128, 64, ensemble_size=3)
        ])
        self.output_layer = nn.Linear(64, 10)

    def forward(self, x):
        x = torch.flatten(x, 1)
        for layer in self.fc_layers:
            x = layer(x)
        return self.output_layer(x)

def GetInput():
    return torch.randn(1, 3, 32, 32).to(torch.float32)

if __name__ == "__main__":
    model = MyModel().to("cuda")
    inp = GetInput().to("cuda")

    # Works without eval()
    model.eval()  # Uncommenting this causes export to fail
    _=model(inp)
    print("Output success")

    new_model = torch.export.export(model, (inp,)).module()
    print("Export success")
    print(new_model(inp))
```
Observed Behavior
When `model.eval()` is called before export, the following error occurs:
```
torch._dynamo.exc.Unsupported: Attempted to call function marked as skipped
  Explanation: Dynamo developers have intentionally marked that the function `typename` in file `/path/to/torch/__init__.py` should not be traced.
  Hint: Avoid calling the function `typename`.
```
The error points to this call in the model's forward path:
```
setattr(self.module, name_w, w)
  -> f"cannot assign '{torch.typename(value)}' as parameter '{name}' "
```

### Versions

```
PyTorch version: 2.7.1a0+gite2d141d
Is debug build: True
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 12.3.0-1ubuntu1~22.04) 12.3.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.16 (main, Dec 11 2024, 16:24:50) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.8.0-59-generic-x86_64-with-glibc2.35
Is CUDA available: True
```

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185335

The aten.dropout decomposition used aten.native_dropout semantics for the no-op branch and returned input.clone() when train was false or p was zero. Eager aten.dropout instead returns the original tensor object for these no-op cases.

That difference matters when Dynamo traces dropout on a Parameter. The cloned FakeTensor no longer preserves the original Parameter identity, so assigning it back through Module.__setattr__ can take the registered-parameter error path and try to format torch.typename, which Dynamo intentionally skips. Return the input directly in the no-op decomposition branch so tracing matches eager dropout aliasing.

A narrower alternative would have been to special-case Module.__setattr__ or torch.typename tracing, but that would only hide the symptom. Matching dropout's eager identity semantics fixes the source of the incorrect FakeTensor identity.

Fixes #160241
Fixes #115249
Generated by my agent

Benchmark Results:
A simple CPU microbenchmark comparing eval dropout identity against the old clone behavior on a 4096x4096 tensor shows the fix removes the clone cost:

- eval_dropout_identity_current: 0.000160s total for 100 iters, 1.60 us/iter
- clone_baseline_old_decomp_behavior: 0.771761s total for 100 iters, 7717.61 us/iter
- speedup_vs_clone_baseline: 4831.36x

Test Plan:
- python test/dynamo/test_repros.py -k test_dropout_eval_setattr_parameter
- python test/dynamo/test_repros.py -k dropout
- minimal torch.compile fullgraph repro from #160241
- minimal torch.export.export repro from #160241
- lintrunner -a
- git diff --cached --check

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Review by ezyang — APPROVED** [review]

**Comment by jansel** [comment]
My agent says CI was failing in test/functorch/test_aotdispatch.py::TestAOTExport::test_aot_export_ban_dropout_mut_pre_dispatch because the test still expected the old clone behavior. I updated the AOT export expectation to assert the corrected no-op dropout alias/input-mutation semantics.

**Comment by jansel** [comment]
My agent says mergebot closed this PR as merged, but the patch is not present on origin/main and GitHub still reports it as unmerged. I reopened the PR and am retrying the merge.

**Comment by jansel** [comment]
@claude explain those CI failures, are they caused by this PR?

**Comment by jansel** [comment]
My agent says mergebot closed this PR again as if it had merged, but the cited commit belongs to a different PR and origin/main still does not contain this patch. I reopened the PR, removed the stale Merged label, and am rebasing/resubmitting the stack.

**Comment by jansel** [comment]
Agent, rebase to origin/main and double check this change has not already landed.  I do see this in the commit on main, but perhaps it has been reverted.  Do you see a revert commit?  Can you figure out why it was reverted?

**Comment by jansel** [comment]
My agent says I rebased and resubmitted this PR onto current PyTorch main (05ce1806e61a). The patch has not already landed: main still has input.clone() in the aten.dropout decomposition, lacks test_dropout_eval_setattr_parameter, and still has the old clone expectation in test_aot_export_ban_dropout_mut_pre_dispatch. I do not see a relevant revert commit. The prior mergebot-cited commits are unrelated: 26296f2e0ecb is #184316 and f5d8a565d81e is #185635. Fresh CI is now running on the resubmitted head.

**Comment by jansel** [comment]
My agent says the latest Dr.CI update shows a single cpp/test_jit failure in the linux-jammy-py3.14-clang18 shard. This PR only changes the Python dropout decomposition/test expectations, so I reran the failed job and will follow up when the retry finishes.

**Comment by jansel** [comment]
@claude was this PR merged to main and reverted?  Or was it never actually merged in the first place?

**Comment by jansel** [comment]
@claude explain those CI failures, are they related to this PR?
